frontend/packages/dev-console/src/components/project-access/project-access-form-utils.ts

```diff
@@ -0,0 +1,166 @@
+    roleBinding.subjects[0].name = user.user;
+    roleBinding.roleRef.name = user.role;
+    roleBinding.metadata.name = roleBindingName;
+    finalArray.push(sendK8sRequest(verb, roleBinding));
+  });
+  Promise.all(finalArray)
+    .then(() => {
+      actions.setSubmitting(false);
+      actions.resetForm({
+        values: {
+          projectAccess: finalRoles,
+        },
+        status: { success: `Successfully updated the ${formName}.` },
+      });
+    })
+    .catch((err) => {
+      actions.setSubmitting(false);
+      actions.setStatus({ submitError: err.message });
+    });
+};
+
+export const validationSchema = yup.object().shape({
+  projectAccess: yup.array().of(
+    yup.object().shape({
+      user: yup.string().required('Required'),
+      role: yup.string().required('Required'),
+    }),
+  ),
+});
+
+export const filterRoleBindings = (roleBindings, roles) => {
+  return _.filter(roleBindings.data, (user: UserRole) => _.keys(roles).includes(user.roleRef.name));
+};
+
+export const getUserRoleBindings = (roleBindings) => {
+  let userRoleBindings: UserRoleBinding[] = [];
+  roleBindings.map(
+    (user: UserRole) =>
+      (userRoleBindings = [
+        ...userRoleBindings,
+        ...[
+          {
+            roleBindingName: user.metadata.name,
+            user: user.subjects[0].name,
+            role: user.roleRef.name,
+          },
+        ],
+      ]),
+  );
+  return userRoleBindings;
+};
+
+export const getNewRoles = (initialRoles: UserRoleBinding[], formValues: UserRoleBinding[]) => {
+  const newRoles = _.uniqBy(
+    _.filter(
+      formValues,
+      (o1) => !initialRoles.find((o2) => o1.user === o2.user && o1.role === o2.role),
+    ),
+    function(user) {
+      return JSON.stringify([user.user, user.role]);
+    },
+  );
+  return newRoles;
+};
+
+export const getRemovedRoles = (initialRoles: UserRoleBinding[], formValues: UserRoleBinding[]) => {
+  const removeRoles = _.filter(
+    initialRoles,
+    (o1) => !formValues.find((o2: UserRoleBinding) => o1.user === o2.user && o1.role === o2.role),
+  );
+  return removeRoles;
+};
+
+export const getRolesWithNameChange = (
+  newRoles: UserRoleBinding[],
+  removeRoles: UserRoleBinding[],
+) => {
+  const createRoles = _.filter(newRoles, 'roleBindingName');
+  const deleteRoles = _.filter(removeRoles, (o1) =>
+    createRoles.find((o2) => o1.roleBindingName === o2.roleBindingName),
+  );
+  const rolesWithNameChange = _.filter(createRoles, (o1) =>
+    deleteRoles.find((o2) => o1.roleBindingName === o2.roleBindingName && o1.user !== o2.user),
+  );
+  return rolesWithNameChange;
+};
+
+export const getFinalRoles = (
+  initialValues: UserRoleBinding[],
+  removeRoles: UserRoleBinding[],
+  newRoles: UserRoleBinding[],
+) => {
+  const finalRoles = _.filter(
+    initialValues,
+    (o1) => !removeRoles.find((o2) => o1.roleBindingName === o2.roleBindingName),
+  );
+  return [...finalRoles, ...newRoles];
+};
```

Comment:
nit: this file has become too big for my liking. Could have split into multiple files like `types`, `submit-utils`, `validation-utils` and `utils`.